src/inausoft.netCLI/netCLI.cs

```diff
@@ -0,0 +1,99 @@
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException($"{nameof(services)} cannot be null.");
+            }
+
+            CLIConfiguration configuration = new CLIConfiguration();
+
+            setup(configuration);
+
+            services.AddSingleton(configuration);
+
+            configuration._commandMap.Values.ToList().ForEach(it =>
+            {
+                services.AddSingleton(it);
+            });
+
+            return services;
+        }
+
+        /// <summary>
+        /// Runs <see cref="ICommandHandler"/> for command specified in args./>
+        /// </summary>
+        /// <param name="serviceProvider"></param>
+        /// <param name="args">Application's command line arguments.</param>
+        /// <returns></returns>
+        public static int RunCLI(this IServiceProvider serviceProvider, string[] args)
+        {
+            var config = serviceProvider.GetService<CLIConfiguration>();
+
+            if (config == null)
+            {
+                throw new InvalidOperationException($"{nameof(CLIConfiguration)} was not registered. Run '{nameof(AddCLI)}' first.");
+            }
+
+            var commandType = config._commandMap.Keys.FirstOrDefault(it =>
+                Attribute.IsDefined(it, typeof(CommandAttribute)) &&
+                (Attribute.GetCustomAttribute(it, typeof(CommandAttribute)) as CommandAttribute).Name == args[0]);
+
+            if (commandType == null)
+            {
+                throw new InvalidCommandException(args[0], $"Command {args[0]} was not mapped.");
+            }
+
+            var command = config.Deserializer.Deserialize(commandType, args.Skip(1).ToArray());
+
+            var handler = serviceProvider.GetRequiredService(config._commandMap[command.GetType()]) as ICommandHandler;
+
+            return handler.Run(command);
+        }
+
+        public static int RunCLI(CLIConfiguration config, string[] args, params ICommandHandler[] handlers)
+        {
+            if (config == null)
+            {
+                throw new InvalidOperationException($"{nameof(CLIConfiguration)} was not registered. Run '{nameof(AddCLI)}' first.");
+            }
+
+            var commandType = config._commandMap.Keys.FirstOrDefault(it =>
+                Attribute.IsDefined(it, typeof(CommandAttribute)) &&
+                (Attribute.GetCustomAttribute(it, typeof(CommandAttribute)) as CommandAttribute).Name == args[0]);
+
+            if (commandType == null)
+            {
+                throw new InvalidCommandException(args[0], $"Command {args[0]} was not mapped.");
+            }
+
+            var command = config.Deserializer.Deserialize(commandType, args.Skip(1).ToArray());
+
+            var handler = handlers.First(it => it.GetType() == config._commandMap[command.GetType()]) as ICommandHandler;
+
+            return handler.Run(command);
```

Comment:
I imagine that `Run` method could return a more complex result object, incl. status code, text to output to the console (currently, how can I verify what was the output for my command? Not sure if I need this though) and maybe some generic data related to executed command. Then you could imagine executing code similar to:
```
var result1 = provider.RunCLI(args1);
...
var result2 = provider.RunCLI(args2); // args2 derived from result1
```
This could allow to use this command+handler pattern not only in CLI apps but maybe in more complex scenarios. Not sure if it's the purpose of this package actually but it seems interesting!

Reply:
Happy to disscus that in next versions - for now I wouldn't go that road.